drivers/common/virtio_mi/lm.c

```diff
@@ -0,0 +1,1177 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2022 NVIDIA Corporation & Affiliates
+ */
+#include <unistd.h>
+#include <net/if.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <rte_malloc.h>
+#include <rte_log.h>
+#include <rte_errno.h>
+#include <rte_string_fns.h>
+#include <rte_bus_pci.h>
+#include <rte_vfio.h>
+#include <rte_kvargs.h>
+#include <rte_eal_paging.h>
+#include <rte_ether.h>
+
+#include <virtqueue.h>
+#include <virtio_admin.h>
+#include <virtio_api.h>
+#include <virtio_lm.h>
+
+#define VIRTIO_VDPA_MI_SUPPORTED_NET_FEATURES (1ULL << VIRTIO_F_ADMIN_VQ)
+
+#define VIRTIO_VDPA_MI_SUPPORTED_BLK_FEATURES (1ULL << VIRTIO_F_ADMIN_VQ)
+
+#define VIRTIO_VDPA_MI_MAX_SGES 32
+
+struct virtio_vdpa_pf_priv;
+struct virtio_vdpa_dev_ops {
+	uint64_t (*get_required_features)(void);
+	uint16_t (*get_adminq_idx)(struct virtio_vdpa_pf_priv *priv);
+};
+
+struct virtio_vdpa_pf_priv {
+	TAILQ_ENTRY(virtio_vdpa_pf_priv) next;
+	struct rte_pci_device *pdev;
+	struct virtio_pci_dev *vpdev;
+	struct virtio_vdpa_dev_ops *dev_ops;
+	uint64_t device_features;
+	int vfio_dev_fd;
+	uint16_t hw_nr_virtqs; /* number of vq device supported*/
+};
+
+struct sge_iova {
```

Comment:
align to spec
diff --git a/content.tex b/content.tex
index c1cb60c..d82d399 100644
--- a/content.tex
+++ b/content.tex
@@ -2116,6 +2116,7 @@ \subsubsection{Dirty page track ctrl}\label{sec:Virtio Transport Options / Virti
struct virtio_sge {
         le64 addr;
         le32 len;
+        le32 reserved;
};